test/distributed/checkpoint/test_fsdp_tp_checkpoint_conversion.py

```diff
@@ -0,0 +1,100 @@
+# Owner(s): ["oncall: distributed"]
+import torch
+import torch.distributed.checkpoint as dist_cp
+from torch.distributed._shard.sharded_tensor import ShardedTensor
+
+from torch.distributed._tensor import DTensor, init_device_mesh, Replicate
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp._shard_utils import _all_gather_sharded_tensor
+from torch.distributed.fsdp.fully_sharded_data_parallel import StateDictType
+
+from torch.distributed.tensor.parallel import PairwiseParallel, parallelize_module
+from torch.testing._internal.common_utils import run_tests
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    MLPModule,
+    skip_if_lt_x_gpu,
+    with_comms,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+
+# TODO: modularize this test and add test for checkpoint conversion in both direction.
+class TestFsdpTpCheckpointConversion(DTensorTestBase):
+    @with_comms
+    @skip_if_lt_x_gpu(2)
+    @with_temp_dir
+    def test_fsdp_to_tp(self):
+        CHECKPOINT_DIR = self.temp_dir
+
+        model = MLPModule(self.device_type).cuda(self.rank)
+        # create a FSDP wrapped model
+        fsdp_model = FSDP(model, use_orig_params=True)
+
+        FSDP.set_state_dict_type(
```

Comment:
this basically means "FSDP use DTensor"?

Reply:
In this example, FSDP is indeed using ShardedTensor, but this is because we did not pass DeviceMesh in FSDP() constructor. 

This `set_state_dict_type` means that we are getting the sharded_state_dict instead of full_state_dict. Sharded_state_dict can be a state_dict with ShardedTensor or DTensor. In this case, it is a ShardedTensor state_dict.

Reply:
ok great, i was also wondering about the devicemesh init being later than FSDP call and you clarified that part.

but if we had initialized FSDP using DeviceMesh, would this example still be relevant for the 2 step sharding discussion for inference?

Reply:
> ok great, i was also wondering about the devicemesh init being later than FSDP call and you clarified that part.
> 
> but if we had initialized FSDP using DeviceMesh, would this example still be relevant for the 2 step sharding discussion for inference?

This unit test is simply to show feasibility for FSDP <-> TP checkpoint conversion on our side. It doesn't cover all the requirements for distributed inference. 

For the 2D inference discussion, I believe they would need to do the following: 
1. Get FairScale checkpoint
2. Run conversion script to convert Fairscale checkpoint to PT-D checkpoint
3. Use DCP to save the converted checkpoint to storage. This is important as directly loading the converted checkpoint into the model may raise a tensor<->DTensor mismatch error. 
4. Create a TP wrapped model 
5. Use DCP to load state_dict from storage
6. Load the state dict into TP wrapped model

Reply:
for step 3 did you mean `loading the **un**converted checkpoint may raise error?

why are step 5 and 6 separate steps?  does 5 put the checkpoint into a separate python dict and then 6 copies the data into the model?

Reply:
> for step 3 did you mean `loading the **un**converted checkpoint may raise error?
> 
> why are step 5 and 6 separate steps? does 5 put the checkpoint into a separate python dict and then 6 copies the data into the model?

Let me clarify the reason for step 3. For step 3, I do mean loading converted checkpoint directly may raise errors. Hamid is using Rohan's conversion script to convert Fairscale checkpoint to PT-D checkpoint. This is to basically convert tensors to ShardedTensor or DTensor. However, Hamid cannot load this directly into TP-lized model because TP does not necessarily put every params into DTensor. Therefore, when Hamid does tp_model.load_state_dict(converted_state_dict), he sees this tensor mismatch error (('aten.copy_.default: got mixed distributed and non-distributed tensors.’,).”). This is the reason he needs to do an extra step of saving the converted state_dict to storage using DCP. DCP does not save the DTensor directly. Instead, it saves tensor with metadata. During loading, DCP requires the model to be initialized first, then it uses storage to load the value. If a parameter is a tensor, DCP would load the value into a tensor. If a parameter is a DTensor, DCP would load the value into a DTensor. This is how it can help resolve the parameter mismatch issue. 

For step 5 and step 6, I just @ you at the actual code pointer in the PR. Step 5 does not put the checkpoint into a separate python dict. Instead, it takes the model.state_dict() from the tp_model before loading the value from storage. 
So, step 5 would be the torch.load() counterpart for DCP, but the difference is DCP operates in place while torch.load() does create a separate python dict. Once we obtain that, we still need to do step 6, calling model.load_state_dict(), no matter whether we are using torch.load() or DCP, as this does copy parameters and buffers from state_dict into this module and its descendants so that training can be resumed. 

Hopefully, this answers your question. Let me know if anything is not clear.